File: rust/python/kimsfinance/strategies/trend.py
"""
Trend-following trading strategies

These strategies use trend indicators like MACD, moving average crossovers,
and trend strength to identify and follow market trends.
"""


class MACDStrategy:
    """
    MACD (Moving Average Convergence Divergence) crossover strategy

    Strategy logic:
    - Buy when MACD line crosses above signal line (bullish crossover)
    - Sell when MACD line crosses below signal line (bearish crossover)

    Parameters:
    - fast_period: Fast EMA period (default: 12)
    - slow_period: Slow EMA period (default: 26)
    - signal_period: Signal line period (default: 9)
    """

    def __init__(self, fast_period=12, slow_period=26, signal_period=9):
        self.fast_period = fast_period
        self.slow_period = slow_period
        self.signal_period = signal_period
        self.prev_macd = None
        self.prev_signal = None

    def on_data(self, bar, indicators):
        """Trading logic with crossover detection"""
        # MACD returns dict with 'macd', 'signal', 'histogram'
        macd = indicators.get('macd_line', 0.0)
        signal = indicators.get('macd_signal', 0.0)

        if self.prev_macd is not None and self.prev_signal is not None:
            # Bullish crossover: MACD crosses above signal
            if self.prev_macd <= self.prev_signal and macd > signal:
                self.prev_macd, self.prev_signal = macd, signal
                return 'buy'
            # Bearish crossover: MACD crosses below signal
            elif self.prev_macd >= self.prev_signal and macd < signal:
                self.prev_macd, self.prev_signal = macd, signal
                return 'sell'

        self.prev_macd, self.prev_signal = macd, signal
        return 'hold'

    def get_indicators(self):
        # Note: Actual implementation would need multi-output indicator support
        return ['macd_line', 'macd_signal']


class EMACrossoverStrategy:
    """
    Exponential Moving Average crossover strategy

    Strategy logic:
    - Buy when fast EMA crosses above slow EMA (golden cross)
    - Sell when fast EMA crosses below slow EMA (death cross)

    Parameters:
    - fast_period: Fast EMA period (default: 12)
    - slow_period: Slow EMA period (default: 26)
    """

    def __init__(self, fast_period=12, slow_period=26):
        self.fast_period = fast_period
        self.slow_period = slow_period
        self.prev_fast = None
        self.prev_slow = None

    def on_data(self, bar, indicators):
        """Trading logic with crossover detection"""
        fast_ema = indicators.get(f'ema_{self.fast_period}', bar['close'])
        slow_ema = indicators.get(f'ema_{self.slow_period}', bar['close'])

        if self.prev_fast is not None and self.prev_slow is not None:
            # Golden cross: fast EMA crosses above slow EMA
            if self.prev_fast <= self.prev_slow and fast_ema > slow_ema:
                self.prev_fast, self.prev_slow = fast_ema, slow_ema
                return 'buy'
            # Death cross: fast EMA crosses below slow EMA
            elif self.prev_fast >= self.prev_slow and fast_ema < slow_ema:
                self.prev_fast, self.prev_slow = fast_ema, slow_ema
                return 'sell'

        self.prev_fast, self.prev_slow = fast_ema, slow_ema
        return 'hold'

    def get_indicators(self):
        return [f'ema_{self.fast_period}', f'ema_{self.slow_period}']


class DualMAStrategy:
    """
    Dual Moving Average strategy with SMA

    Strategy logic:
    - Buy when fast SMA crosses above slow SMA
    - Sell when fast SMA crosses below slow SMA

    Parameters:
    - fast_period: Fast SMA period (default: 50)
    - slow_period: Slow SMA period (default: 200)
    """

    def __init__(self, fast_period=50, slow_period=200):
        self.fast_period = fast_period
        self.slow_period = slow_period
        self.prev_fast = None
        self.prev_slow = None

    def on_data(self, bar, indicators):
        """Trading logic with crossover detection"""
        fast_sma = indicators.get(f'sma_{self.fast_period}', bar['close'])
        slow_sma = indicators.get(f'sma_{self.slow_period}', bar['close'])

        if self.prev_fast is not None and self.prev_slow is not None:
            # Bullish crossover
            if self.prev_fast <= self.prev_slow and fast_sma > slow_sma:
                self.prev_fast, self.prev_slow = fast_sma, slow_sma
                return 'buy'
            # Bearish crossover
            elif self.prev_fast >= self.prev_slow and fast_sma < slow_sma:
                self.prev_fast, self.prev_slow = fast_sma, slow_sma
                return 'sell'

        self.prev_fast, self.prev_slow = fast_sma, slow_sma
        return 'hold'

    def get_indicators(self):
        return [f'sma_{self.fast_period}', f'sma_{self.slow_period}']


class TrendFollowingStrategy:
    """
    Multi-timeframe trend following strategy

    Strategy logic:
    - Enter long when price > long-term MA AND short-term MA slopes up
    - Exit when short-term MA slopes down
    - Uses ATR for position sizing

    Parameters:
    - short_period: Short-term trend detection (default: 20)
    - long_period: Long-term trend filter (default: 50)
    - atr_period: ATR period for position sizing (default: 14)
    - atr_multiplier: Risk per trade in ATR units (default: 2.0)
    """

    def __init__(self, short_period=20, long_period=50, atr_period=14, atr_multiplier=2.0):
        self.short_period = short_period
        self.long_period = long_period
        self.atr_period = atr_period
        self.atr_multiplier = atr_multiplier
        self.prev_short_ma = None

    def on_data(self, bar, indicators):
        """Trading logic with trend detection"""
        short_ma = indicators.get(f'ema_{self.short_period}', bar['close'])
        long_ma = indicators.get(f'ema_{self.long_period}', bar['close'])
        price = bar['close']

        # Check if price is above long-term trend
        in_uptrend = price > long_ma

        if self.prev_short_ma is not None:
            short_ma_slope = short_ma - self.prev_short_ma

            # Buy signal: uptrend + short MA slopes up
            if in_uptrend and short_ma_slope > 0 and self.prev_short_ma is not None:
                if self.prev_short_ma <= price and short_ma > price:  # Price crosses above short MA
                    self.prev_short_ma = short_ma
                    return 'buy'

            # Sell signal: short MA slopes down
            elif short_ma_slope < 0:
                if self.prev_short_ma >= price and short_ma < price:  # Price crosses below short MA
                    self.prev_short_ma = short_ma
                    return 'sell'

        self.prev_short_ma = short_ma
        return 'hold'

    def get_indicators(self):
        return [
            f'ema_{self.short_period}',
            f'ema_{self.long_period}',
            f'atr_{self.atr_period}'
        ]

    def position_size(self, equity, signal):
        """ATR-based position sizing"""
        # In production, would calculate: equity / (atr * atr_multiplier)
        # For now, return fixed allocation
        return 1.0

File: rust/src/strategies/trend.rs
//! Trend-following trading strategies
//!
//! Strategies that identify and follow market trends using moving averages
//! and channel breakouts.

use crate::backtest::core::{
    IndicatorConfig, IndicatorValues, OHLCVBar, ParameterGrid, ParameterRange, Signal, Strategy,
};

// ============================================================================
// EMA Crossover Strategy
// ============================================================================

/// EMA Crossover Strategy (Golden Cross/Death Cross)
///
/// Classic trend-following strategy using fast and slow EMA crossovers.
/// Buys when fast EMA crosses above slow EMA (Golden Cross).
/// Sells when fast EMA crosses below slow EMA (Death Cross).
///
/// # Default Parameters
/// - Fast EMA: 50 (short-term trend)
/// - Slow EMA: 200 (long-term trend)
///
/// # Optimization Ranges
/// - Fast EMA: 20-100
/// - Slow EMA: 100-250
///
/// # Market Conditions
/// - Best: Strong trending markets (bull or bear)
/// - Avoid: Choppy, sideways markets (whipsaws)
///
/// # Risk Management
/// - Stop Loss: 3% below entry
/// - Take Profit: 20% above entry (ride the trend)
#[derive(Debug, Clone)]
pub struct EMACrossover {
    pub fast_period: usize,
    pub slow_period: usize,
    pub stop_loss_pct: f64,
    pub take_profit_pct: f64,
    initial_capital: f64,
    prev_fast_ema: f64,
    prev_slow_ema: f64,
}

impl EMACrossover {
    pub fn new(fast_period: usize, slow_period: usize) -> Self {
        Self {
            fast_period,
            slow_period,
            stop_loss_pct: 0.03,
            take_profit_pct: 0.20,
            initial_capital: 10_000.0,
            prev_fast_ema: 0.0,
            prev_slow_ema: 0.0,
        }
    }
}

impl Default for EMACrossover {
    fn default() -> Self {
        Self::new(50, 200)
    }
}

impl Strategy for EMACrossover {
    fn on_data(&mut self, _bar: &OHLCVBar, indicators: &IndicatorValues) -> Signal {
        let fast_key = format!("ema_{}", self.fast_period);
        let slow_key = format!("ema_{}", self.slow_period);

        let fast_ema = indicators.get(&fast_key).copied().unwrap_or(0.0);
        let slow_ema = indicators.get(&slow_key).copied().unwrap_or(0.0);

        let result = if fast_ema > slow_ema && self.prev_fast_ema <= self.prev_slow_ema {
            Signal::Buy
        } else if fast_ema < slow_ema && self.prev_fast_ema >= self.prev_slow_ema {
            Signal::Sell
        } else {
            Signal::Hold
        };

        self.prev_fast_ema = fast_ema;
        self.prev_slow_ema = slow_ema;

        result
    }

    fn indicators(&self) -> Vec<IndicatorConfig> {
        vec![
            IndicatorConfig::EMA { period: self.fast_period },
            IndicatorConfig::EMA { period: self.slow_period },
        ]
    }

    fn parameters(&self) -> ParameterGrid {
        let mut grid = ParameterGrid::new();
        grid.add_range(
            "fast_period",
            ParameterRange::Int { min: 20, max: 100, step: 10 },
        );
        grid.add_range(
            "slow_period",
            ParameterRange::Int { min: 100, max: 250, step: 25 },
        );
        grid
    }

    fn initial_capital(&self) -> f64 {
        self.initial_capital
    }
}

// ============================================================================
// Triple EMA Trend Strategy
// ============================================================================

/// Triple EMA Trend Strategy
///
/// Uses three EMAs to identify trend strength and direction.
/// Buys when short > medium > long (strong uptrend).
/// Sells when short < medium < long (strong downtrend).
///
/// # Default Parameters
/// - Short EMA: 8 (immediate trend)
/// - Medium EMA: 21 (intermediate trend)
/// - Long EMA: 55 (major trend)
///
/// # Optimization Ranges
/// - Short EMA: 5-15
/// - Medium EMA: 15-30
/// - Long EMA: 40-100
///
/// # Market Conditions
/// - Best: Clear trending markets with momentum
/// - Avoid: Ranging markets with frequent crossovers
///
/// # Risk Management
/// - Stop Loss: 4% below entry
/// - Take Profit: 18% above entry
#[derive(Debug, Clone)]
pub struct TripleEMATrend {
    pub short_period: usize,
    pub medium_period: usize,
    pub long_period: usize,
    pub stop_loss_pct: f64,
    pub take_profit_pct: f64,
    initial_capital: f64,
}

impl TripleEMATrend {
    pub fn new(short_period: usize, medium_period: usize, long_period: usize) -> Self {
        Self {
            short_period,
            medium_period,
            long_period,
            stop_loss_pct: 0.04,
            take_profit_pct: 0.18,
            initial_capital: 10_000.0,
        }
    }
}

impl Default for TripleEMATrend {
    fn default() -> Self {
        Self::new(8, 21, 55)
    }
}

impl Strategy for TripleEMATrend {
    fn on_data(&mut self, _bar: &OHLCVBar, indicators: &IndicatorValues) -> Signal {
        let short_key = format!("ema_{}", self.short_period);
        let medium_key = format!("ema_{}", self.medium_period);
        let long_key = format!("ema_{}", self.long_period);

        let short_ema = indicators.get(&short_key).copied().unwrap_or(0.0);
        let medium_ema = indicators.get(&medium_key).copied().unwrap_or(0.0);
        let long_ema = indicators.get(&long_key).copied().unwrap_or(0.0);

        if short_ema > medium_ema && medium_ema > long_ema {
            Signal::Buy
        } else if short_ema < medium_ema && medium_ema < long_ema {
            Signal::Sell
        } else {
            Signal::Hold
        }
    }

    fn indicators(&self) -> Vec<IndicatorConfig> {
        vec![
            IndicatorConfig::EMA { period: self.short_period },
            IndicatorConfig::EMA { period: self.medium_period },
            IndicatorConfig::EMA { period: self.long_period },
        ]
    }

    fn parameters(&self) -> ParameterGrid {
        let mut grid = ParameterGrid::new();
        grid.add_range(
            "short_period",
            ParameterRange::Int { min: 5, max: 15, step: 2 },
        );
        grid.add_range(
            "medium_period",
            ParameterRange::Int { min: 15, max: 30, step: 5 },
        );
        grid.add_range(
            "long_period",
            ParameterRange::Int { min: 40, max: 100, step: 10 },
        );
        grid
    }

    fn initial_capital(&self) -> f64 {
        self.initial_capital
    }
}

// ============================================================================
// Donchian Channel Breakout Strategy
// ============================================================================

/// Donchian Channel Breakout Strategy (Turtle Trading)
///
/// Classic turtle trading strategy using Donchian channels.
/// Buys on breakout above upper channel (new 20-day high).
/// Sells on breakout below lower channel (new 20-day low).
///
/// # Default Parameters
/// - Channel Period: 20 (turtle standard)
/// - Breakout Confirmation: Price must close above/below channel
///
/// # Optimization Ranges
/// - Channel Period: 10-40
///
/// # Market Conditions
/// - Best: Markets breaking out of consolidation
/// - Avoid: Range-bound markets (false breakouts)
///
/// # Risk Management
/// - Stop Loss: 2% below entry (tight for breakouts)
/// - Take Profit: 25% above entry (capture full breakout)
#[derive(Debug, Clone)]
pub struct DonchianBreakout {
    pub channel_period: usize,
    pub stop_loss_pct: f64,
    pub take_profit_pct: f64,
    initial_capital: f64,
}

impl DonchianBreakout {
    pub fn new(channel_period: usize) -> Self {
        Self {
            channel_period,
            stop_loss_pct: 0.02,
            take_profit_pct: 0.25,
            initial_capital: 10_000.0,
        }
    }
}

impl Default for DonchianBreakout {
    fn default() -> Self {
        Self::new(20)
    }
}

impl Strategy for DonchianBreakout {
    fn on_data(&mut self, bar: &OHLCVBar, indicators: &IndicatorValues) -> Signal {
        let key = format!("dc_{}", self.channel_period);
        let upper_key = format!("{}_upper", key);
        let lower_key = format!("{}_lower", key);

        let upper_channel = indicators.get(&upper_key).copied().unwrap_or(bar.high);
        let lower_channel = indicators.get(&lower_key).copied().unwrap_or(bar.low);

        if bar.close > upper_channel {
            Signal::Buy
        } else if bar.close < lower_channel {
            Signal::Sell
        } else {
            Signal::Hold
        }
    }

    fn indicators(&self) -> Vec<IndicatorConfig> {
        vec![]
    }

    fn parameters(&self) -> ParameterGrid {
        let mut grid = ParameterGrid::new();
        grid.add_range(
            "channel_period",
            ParameterRange::Int { min: 10, max: 40, step: 5 },
        );
        grid
    }

    fn initial_capital(&self) -> f64 {
        self.initial_capital
    }
}

// ============================================================================
// Keltner Channel Trend Strategy
// ============================================================================

/// Keltner Channel Trend Strategy
///
/// Uses Keltner Channels (EMA + ATR) to identify trend strength.
/// Buys when price breaks above upper channel (strong uptrend).
/// Sells when price breaks below lower channel (strong downtrend).
///
/// # Default Parameters
/// - EMA Period: 20 (middle line)
/// - ATR Period: 10 (channel width)
/// - ATR Multiplier: 2.0 (channel distance)
///
/// # Optimization Ranges
/// - EMA Period: 10-30
/// - ATR Period: 5-20
/// - ATR Multiplier: 1.5-3.0
///
/// # Market Conditions
/// - Best: Trending markets with expanding volatility
/// - Avoid: Low volatility consolidation
///
/// # Risk Management
/// - Stop Loss: ATR-based (1.5x ATR below entry)
/// - Take Profit: 15% above entry
#[derive(Debug, Clone)]
pub struct KeltnerTrend {
    pub ema_period: usize,
    pub atr_period: usize,
    pub atr_multiplier: f64,
    pub stop_loss_atr_mult: f64,
    pub take_profit_pct: f64,
    initial_capital: f64,
}

impl KeltnerTrend {
    pub fn new(ema_period: usize, atr_period: usize, atr_multiplier: f64) -> Self {
        Self {
            ema_period,
            atr_period,
            atr_multiplier,
            stop_loss_atr_mult: 1.5,
            take_profit_pct: 0.15,
            initial_capital: 10_000.0,
        }
    }
}

impl Default for KeltnerTrend {
    fn default() -> Self {
        Self::new(20, 10, 2.0)
    }
}

impl Strategy for KeltnerTrend {
    fn on_data(&mut self, bar: &OHLCVBar, indicators: &IndicatorValues) -> Signal {
        let key = format!("kc_{}_{}", self.ema_period, self.atr_period);
        let upper_key = format!("{}_upper", key);
        let lower_key = format!("{}_lower", key);

        let upper_channel = indicators.get(&upper_key).copied().unwrap_or(bar.high);
        let lower_channel = indicators.get(&lower_key).copied().unwrap_or(bar.low);

        if bar.close > upper_channel {
            Signal::Buy
        } else if bar.close < lower_channel {
            Signal::Sell
        } else {
            Signal::Hold
        }
    }

    fn indicators(&self) -> Vec<IndicatorConfig> {
        vec![]
    }

    fn parameters(&self) -> ParameterGrid {
        let mut grid = ParameterGrid::new();
        grid.add_range(
            "ema_period",
            ParameterRange::Int { min: 10, max: 30, step: 5 },
        );
        grid.add_range(
            "atr_period",
            ParameterRange::Int { min: 5, max: 20, step: 5 },
        );
        grid.add_range(
            "atr_multiplier",
            ParameterRange::Float { min: 1.5, max: 3.0, step: 0.5 },
        );
        grid
    }

    fn initial_capital(&self) -> f64 {
        self.initial_capital
    }
}
